### ISSUE TITLE
useSortBy should toggle`aria-sort` on column headers and relevant demos should encourage a11y best practices

### ISSUE THREAD

**Issue by nicklemmon** [TanStack/react-table#2992]
**Describe the bug (required)**

1. `useSortBy` should leverage the [`aria-sort`](https://www.w3.org/TR/wai-aria-1.1/#aria-sort) attribute which describes the current sort direction for a particular column to screen reader only users
2. [The demo for `useSortBy` is not keyboard accessible](https://codesandbox.io/s/github/tannerlinsley/react-table/tree/master/examples/sorting?file=/src/App.js) - fortunately this can be overcome easily via manually leveraging `column.toggleSortBy`! Since this library is fairly popular, it would be good to encourage best practices where possible.

**Provide an example via Codesandbox! (required)**
[Codesandbox Demo](https://codesandbox.io/s/react-table-sorting-a11y-problems-gyu12) <-- here is a demo demonstrating the lack of `aria-sort` on the `<th>` elements _with_ a keyboard accessible implementation using `column.toggleSortBy`

**Steps To Reproduce (required)**

1. Go to the [linked demo](https://codesandbox.io/s/github/tannerlinsley/react-table/tree/master/examples/sorting) in the documentation for `useSortBy`
2. Try to control the sorting using a keyboard only - unfortunately you can't due to some HTML semantics problems
3. When clicking on a `<th>` to sort, inspect the DOM for the `aria-sort` attribute - it isn't set at all.

**Expected behavior (Recommended)**
Here is a great demo of how this should work from a reliable source:
https://dequeuniversity.com/library/aria/table-sortable

1. All sortable `<th>` elements have a nested `<button>` inside to make them keyboard accessible
2. All sortable `<th>` elements have a valid `aria-sort` attribute (but only if they are sortable!)
3. Note they do _not_ have the `title` attribute that is currently provided by `useSortBy` - this shouldn't be necessary with the `aria-sort` attribute in place.

**Screenshots**
N/A

**Desktop (please complete the following information):**

- OS: macOS Big Sur
- Browser: All browsers
- Version: 7.6.3

**Smartphone (please complete the following information):**

N/A

**Additional context**
Seriously, thanks to all of the maintainers for your work. The headless approach this library takes serves the React community well.

Depending on the complexity of the code, I could potentially jump in and help out with this one. The `aria-sort` problem in particular seems like low-hanging fruit. Updating a CodeSandbox demo should be quite easy as well, though I'm not sure of the best way to go about doing that in this case.

**Comment by hakubo** [comment]
In addition to that I have noticed that it sets title on header cell to "Toggle SortBy". I think name of the column should be preserved here like: 'Toggle SortBy, column: ${name}"

**Comment by nicklemmon** [comment]
> In addition to that I have noticed that it sets title on header cell to "Toggle SortBy". I think name of the column should be preserved here like: 'Toggle SortBy, column: ${name}"

Right - definitely not the right content. I think when paired with aria-sort, the regular heading content should do just fine for screen reader users.

**Comment by esetnik** [comment]
@tannerlinsley I believe this should be reopened as it's still an issue with react-table v8. It makes it hard to target sortable headers in cypress tests in an accessible way.

**Comment by tannerlinsley** [comment]
It’s headless, so you should be able to put your own area attribute in your market depending on the table state

**Comment by esetnik** [comment]
For anyone else that comes across this. I made it work using the following:

```tsx
                <TableHeaderColumn
                  {...column.getHeaderProps(column.getSortByToggleProps())}
                  aria-sort={
                    column.isSorted
                      ? column.isSortedDesc
                        ? 'descending'
                        : 'ascending'
                      : 'none'
                  }
                >
                  <TableHeaderCell role="button">
                    {column.render('Header')}
                    {column.canSort && (
                      <SortIcon
                        size={IconSize.extraSmall}
                        direction={
                          column.isSorted
                            ? column.isSortedDesc
                              ? Direction.down
                              : Direction.up
                            : null
                        }
                      />
                    )}
                  </TableHeaderCell>
                </TableHeaderColumn>
```

**Comment by nicklemmon** [comment]
Nice, @esetnik!

A couple of things to be sure to address too (your component might do this under the hood):

* Make sure the rendered `th` does *not* have `role="button"` - that would wipe out the semantic meaning of the `th` and make `aria-sort` meaningless
* The rendered markup should render like so at the end of the day:

```html
<th aria-sort="descending">
  <button>Table header content here</button>
</th>
```

Check out this example for more:
https://dequeuniversity.com/library/aria/table-sortable

**Comment by nicklemmon** [comment]
And @tannerlinsley in response to your comment here:

> It’s headless, so you should be able to put your own area attribute in your market depending on the table state

It _is_ headless though it does render relevant HTML attributes for some features of the table but not others. `col-span` and `row-span`, for example:
https://tanstack.com/table/v8/docs/api/core/header#colspan

I suppose it depends on what "headless" means exactly! For some headless libraries (React ARIA, Downshift, Radix) they are all about adding the right HTML attributes for handling state changes (including ARIA). Is this library only concerned with data? If it's concerned at all with producing the right HTML to reflect table state, then ARIA attributes should be part of the mix I would think.

(And thanks for all your hard work on this and so many other libraries 🙏  - big fan of your work)

**Comment by esetnik** [comment]
> Nice, @esetnik!
> 
> A couple of things to be sure to address too (your component might do this under the hood):
> 
> * Make sure the rendered `th` does _not_ have `role="button"` - that would wipe out the semantic meaning of the `th` and make `aria-sort` meaningless
> * The rendered markup should render like so at the end of the day:
> 
> ```
> <th aria-sort="descending">
>   <button>Table header content here</button>
> </th>
> ```
> 
> Check out this example for more: https://dequeuniversity.com/library/aria/table-sortable

```html
<th colspan="1" role="columnheader" title="Toggle SortBy" aria-sort="ascending" class="Table__TableHeaderColumn-sc-13lnlsj-5 jtfJwa" style="cursor: pointer;"><div role="button" aria-pressed="true" class="Table__TableHeaderCell-sc-13lnlsj-0 ejdZOg">Class Name<div size="1" direction="0" color="black" class="Icons__BaseIcon-sc-11pp0d5-0 Icons__Arrow-sc-11pp0d5-54 Icons__SortIcon-sc-11pp0d5-56 hQXkhp ilQtJx kiMnuM"></div></div></th>
```

**Comment by esetnik** [comment]
> I suppose it depends on what "headless" means exactly! For some headless libraries (React ARIA, Downshift, Radix) they are all about adding the right HTML attributes for handling state changes (including ARIA). Is this library only concerned with data? If it's concerned at all with producing the right HTML to reflect table state, then ARIA attributes should be part of the mix I would think.

I agree. And most devs don't know too much about `aria` and often don't have the business justification to bother supporting it. In my opinion it should be baked in whenever possible. If adding too many properties by default was a concern there could be some configuration one could set to decide whether or not to generate aria attributes.

**Comment by tannerlinsley** [comment]
I’m happy to accept any PRs for better aria attrs!
